src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -146,21 +148,31 @@ private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateA
                     continue;
                 }
 
-                if (DiagnosticsAreUnchanged(documentToPreviousDiagnosticParams, document))
-                {
-                    context.TraceInformation($"Diagnostics were unchanged for document: {document.FilePath}");
+                TReport report;
 
-                    // Nothing changed between the last request and this one.  Report a (null-diagnostics,
-                    // same-result-id) response to the client as that means they should just preserve the current
-                    // diagnostics they have for this file.
-                    var previousParams = documentToPreviousDiagnosticParams[document];
-                    progress.Report(CreateReport(previousParams.TextDocument, diagnostics: null, previousParams.PreviousResultId));
-                }
-                else
+                // Ensure the read to _documentIdToLastResultId to get the cached resultId and subsequent write
+                // to _documentIdToLastResultId after calculating new diagnostics happens in a single 'transaction'
+                // to prevent us from overwriting data in _documentIdToLastResultId with stale results.
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
                 {
-                    context.TraceInformation($"Diagnostics were changed for document: {document.FilePath}");
-                    await ComputeAndReportCurrentDiagnosticsAsync(context, progress, document, cancellationToken).ConfigureAwait(false);
+                    if (DiagnosticsAreUnchanged(documentToPreviousDiagnosticParams, document))
+                    {
+                        context.TraceInformation($"Diagnostics were unchanged for document: {document.FilePath}");
+
+                        // Nothing changed between the last request and this one.  Report a (null-diagnostics,
+                        // same-result-id) response to the client as that means they should just preserve the current
+                        // diagnostics they have for this file.
+                        var previousParams = documentToPreviousDiagnosticParams[document];
+                        report = CreateReport(previousParams.TextDocument, diagnostics: null, previousParams.PreviousResultId);
+                    }
+                    else
+                    {
+                        context.TraceInformation($"Diagnostics were changed for document: {document.FilePath}");
+                        report = await ComputeAndReportCurrentDiagnosticsAsync(context, document, cancellationToken).ConfigureAwait(false);
```

Comment:
having a lock around CmoputeAndReport is definitely scary.  computation can take unbounded time.  is it possible we can still compute outside of hte lock, but orderr things to always be consistent with teh results?

Reply:
yeah that make sense,, it could block us hearing about diagnostic updates for quite a while.

I wonder if I could update the _documentIdToLastResultId map _before_ I do the computation instead of after, then release the lock and allow computation to proceed.  Could potentially do that inside a try and clear out _documentIdToLastResultId if the computation throws an exception for some reason.

Reply:
discussed offline, Moved caching before calculation as we will only report resultIds (per document) to client if the calculation succeeds.  We do not report partial document results.  If there is a cancellation or exception, the client will not see a resultId for the document at all, and therefore cannot ask us for it.